File: VirtualLab_Python/SynDiag_Python/diagnostics/Tokalab/Diag_SaddleCoils.py
# -*- coding: utf-8 -*-
"""
Created on Tue Jun  3 17:05:34 2025

@author: ricca
"""

import numpy as np
from scipy.interpolate import griddata
import os
import matplotlib.pyplot as plt

class Diag_SaddleCoils:
    
    def __init__(self):
        self.R1 = None
        self.Z1 = None
        self.R2 = None
        self.Z2 = None
        self.Dpsi = None
        self.sigma_Dpsi = None
        self.unit = None
        self.config = {}
        self.ideal = {}

    def measure(self, equi):
        # Estrai la griglia e il campo psi
        R_equi = equi.geo.grid.Rg
        Z_equi = equi.geo.grid.Zg
        psi_equi = equi.psi

        # Appiattisci le coordinate per griddata
        points = np.column_stack((R_equi.flatten(), Z_equi.flatten()))
        values = psi_equi.flatten()

        # Interpola psi nei due punti
        psi1 = griddata(points, values, (self.R1, self.Z1), method='linear')
        psi2 = griddata(points, values, (self.R2, self.Z2), method='linear')

        # Calcola differenza
        Dpsi = psi2 - psi1

        self.ideal['Dpsi'] = Dpsi

        # Rumore assoluto
        noise_abs = np.random.normal(
            0,
            self.config.get('noise_random_absolute_intensity', 0),
            size=np.shape(Dpsi)
        )

        # Rumore proporzionale
        noise_prop = np.random.normal(
            0,
            np.abs(Dpsi) * self.config.get('noise_random_proportional_intensity', 0)
        )

        # Aggiungi il rumore alla misura
        self.Dpsi = Dpsi + noise_abs + noise_prop
        self.sigma_Dpsi = np.sqrt(self.config.get('noise_random_absolute_intensity')**2 +
                                  (np.abs(Dpsi) * self.config.get('noise_random_proportional_intensity'))**2)
        self.unit = "Wb/rad"

        return self

    def upload(self, configuration=1):
        if configuration == 1:
            self.config['configuration'] = 1

            # Percorso del file .mat (adatta se necessario)
            from scipy.io import loadmat
            module_path = os.path.abspath(__file__)
            module_path = os.path.dirname(module_path)
            module_path = os.path.join(module_path,"diagnostics_data","SaddleCoilsData_config_1.mat")
            data = loadmat(module_path)

            # Carica i dati
            self.R1 = data['R1']
            self.Z1 = data['Z1']
            self.R2 = data['R2']
            self.Z2 = data['Z2']

            self.config['noise_random_absolute_intensity'] = 0
            self.config['noise_random_proportional_intensity'] = 0
            
        elif configuration == 2:
            
            from scipy.io import loadmat
            module_path = os.path.abspath(__file__)
            module_path = os.path.dirname(module_path)
            module_path = os.path.join(module_path,"diagnostics_data","SaddleCoilsData_config_1.mat")
            data = loadmat(module_path)

            # Carica i dati
            self.R1 = data['R1']
            self.Z1 = data['Z1']
            self.R2 = data['R2']
            self.Z2 = data['Z2']

            self.config['noise_random_absolute_intensity'] = 0.1
            self.config['noise_random_proportional_intensity'] = 0.1

        return self
    
    #### Plotting functions
    
    def plot_geo(self):
        R = np.concatenate([self.R1, self.R2])
        Z = np.concatenate([self.Z1, self.Z2])
    
        plt.plot(R, Z, '.-b', markersize=16, linewidth=1.2)
    
        plt.grid(visible=True, which='both')
        plt.xlabel("R")
        plt.ylabel("Z")
        plt.show()
        
    def plot_meas(self):
        # Plot Dpsi values as dots
        plt.plot(self.Dpsi.T, '.', markersize=16)
    
        plt.grid(visible=True, which='both')
        plt.xlabel("#")
        plt.ylabel(r"$\psi$ [Wb/rad]")
        plt.show()
    
    def plot_StandAlone(self):
        # Clear previous plot
        plt.clf()
    
        # Plot ideal Dpsi values as blue dots
        plt.plot(self.ideal['Dpsi'].T, '.b', markersize=16)
    
        # Plot actual Dpsi values as red circles
        plt.plot(self.Dpsi.T, 'or', linewidth=1.2)
    
        plt.grid(visible=True, which='both')
        plt.xlabel("#")
        plt.ylabel(r"$\psi$ [Wb/rad]")
        plt.show()

File: VirtualLab_MatLab/SynDiag_MATLAB/diagnostics/Diag_SaddleCoils.m
%% tokalab diagnostics

classdef Diag_SaddleCoils

    properties

        R1 % Horixontal coordinate
        Z1 % Vertical coordinate

        R2 % Horixontal coordinate
        Z2 % Vertical coordinate

        Dpsi % Measured Flux

        sigma_Dpsi % Associated Uncertainty to Measured Flux

        unit % Unit Measure

        config % contains the various information such as noise, etc.

        ideal % contains the measurements without the noise

    end

    methods

        function obj = measure(obj,equi)

            R_equi = equi.geo.grid.Rg;
            Z_equi = equi.geo.grid.Zg;

            psi_equi = equi.psi;

            Dpsi = interp2(R_equi,Z_equi,psi_equi,obj.R2,obj.Z2) -...
                interp2(R_equi,Z_equi,psi_equi,obj.R1,obj.Z1);

            obj.ideal.Dpsi = Dpsi;

            % noise absolute
            noise_abs = normrnd(0,obj.config.noise_random_absolute_intensity,size(obj.ideal.Dpsi));

            % noise proportional
            noise_prop = normrnd(0,abs(obj.ideal.Dpsi).*obj.config.noise_random_proportional_intensity);

            % real measurement
            obj.Dpsi = obj.ideal.Dpsi + noise_abs + noise_prop;

            % associated uncertainty
            obj.sigma_Dpsi = sqrt(obj.config.noise_random_absolute_intensity.^2 + ...
                (abs(obj.ideal.Dpsi).*obj.config.noise_random_proportional_intensity).^2);

            obj.unit = "Wb/rad";

        end

        function obj = Upload(obj,configuration,machine)

            % default configuration
            if nargin<2
                configuration = 1;
                machine = "TokaLab";
            elseif nargin < 3
                machine = "TokaLab";
            end

            %%%%%%%%%%%% TokaLab Configuration %%%%%%%%%%%%%%%%%%%%%%%%%%%%
            if machine == "TokaLab"
                if configuration == 1

                    obj.config.configuration = 1;

                    load("SaddleCoilsData_TokaLab_config_1.mat")

                    obj.R1 = R1;
                    obj.Z1 = Z1;

                    obj.R2 = R2;
                    obj.Z2 = Z2;

                    obj.config.noise_random_absolute_intensity = 0;

                    obj.config.noise_random_proportional_intensity = 0;

                end
            end

            %%%%%%%%%%%% JET-like Configuration %%%%%%%%%%%%%%%%%%%%%%%%%%%%

            if machine == "JET-like"
                if configuration == 1

                    obj.config.configuration = 1;

                    load("SaddleCoilsData_JETlike_config_1.mat")

                    obj.R1 = R1;
                    obj.Z1 = Z1;

                    obj.R2 = R2;
                    obj.Z2 = Z2;

                    obj.config.noise_random_absolute_intensity = 0;

                    obj.config.noise_random_proportional_intensity = 0;

                end
            end

            %%%%%%%%%%%% DTT-like Configuration %%%%%%%%%%%%%%%%%%%%%%%%%%%%

            if machine == "DTT-like"
                if configuration == 1

                    obj.config.configuration = 1;

                    load("SaddleCoilsData_DTTlike_config_1.mat")

                    obj.R1 = R1;
                    obj.Z1 = Z1;

                    obj.R2 = R2;
                    obj.Z2 = Z2;

                    obj.config.noise_random_absolute_intensity = 0;

                    obj.config.noise_random_proportional_intensity = 0;

                end
            end

        end

        %% Plotting Functions

        function plot_geo(obj)

            plot([obj.R1; obj.R2],[obj.Z1; obj.Z2],...
                '.-b','MarkerSize',16,'LineWidth',1.2)
            grid on
            grid minor
            xlabel("R")
            ylabel("Z")

        end

        function plot_meas(obj)

            plot(obj.Dpsi,'.','MarkerSize',16)
            grid on
            grid minor
            xlabel("#")
            ylabel("\psi [Wb/rad]")

        end

        function plot_StandAlone(obj)

            hold off
            plot(obj.ideal.Dpsi,'.b','MarkerSize',16)
            hold on
            plot(obj.Dpsi,'or','LineWidth',1.2)
            grid on
            grid minor
            xlabel("#")
            ylabel("\psi [Wb/rad]")

        end

    end

end




